File: services/analytics/app/offer_generator.py
from decimal import Decimal

from app.schemas import DealStrategy, OfferLetterRequest, OfferLetterResponse

DISCLAIMER = (
    "BOZZA GENERATA AUTOMATICAMENTE — Non costituisce consulenza legale o fiscale. "
    "Revisione obbligatoria da parte di avvocato/notaio prima dell'invio."
)


def generate_offer_letter(req: OfferLetterRequest) -> OfferLetterResponse:
    discount = (
        ((req.asking_price - req.offered_price) / req.asking_price * Decimal("100"))
        if req.asking_price > 0
        else Decimal("0")
    )
    deposit = req.offered_price * req.deposit_pct

    strategy_labels = {
        DealStrategy.FIX_FLIP: "acquisto con finalità di riqualificazione e successiva vendita",
        DealStrategy.BUY_RENOVATE_RENT: "acquisto con ristrutturazione e locazione",
        DealStrategy.BUY_HOLD_SELL: "acquisto con gestione patrimoniale e exit programmata",
    }

    points_text = "\n".join(f"  • {p}" for p in req.key_points) if req.key_points else (
        "  • Prezzo coerente con analisi di mercato e stato dell'immobile\n"
        "  • Tempistiche definite per perizia e accesso tecnico"
    )

    docs_text = "\n".join(f"  • {d}" for d in req.requested_documents)

    commercial = f"""PROPOSTA DI ACQUISTO — BOZZA COMMERCIALE

Oggetto: Immobile sito in {req.property_address}

Gentile Venditore,

con la presente formuliamo proposta irrevocabile per {req.closing_days} giorni
per l'acquisto dell'immobile in oggetto, alle seguenti condizioni economiche:

  Prezzo richiesto:     € {req.asking_price:,.2f}
  Prezzo offerto:       € {req.offered_price:,.2f} (sconto circa {discount:.1f}%)
  Caparra confirmatoria: € {deposit:,.2f} ({req.deposit_pct * 100:.0f}% del prezzo offerto)

Finalità dell'acquisto: {strategy_labels.get(req.strategy, 'investimento immobiliare')}.

Motivazione sintetica:
{points_text}

Documentazione richiesta prima del rogito:
{docs_text}

Tempistiche proposte:
  • Accettazione proposta: entro {req.closing_days // 3} giorni
  • Perizia tecnica e verifiche documentali: entro {req.closing_days // 2} giorni
  • Rogito: entro {req.closing_days} giorni dall'accettazione

Restiamo a disposizione per un sopralluogo tecnico e per approfondire
ogni aspetto dell'operazione.

Cordiali saluti,
[SOCIETÀ ACQUIRENTE — da compilare]
"""

    placeholders = [
        {
            "key": "CLAUSOLA_SOSPENSIVA",
            "text": "[INSERIRE clausole sospensive: mutuo, conformità urbanistica, assenza vincoli]",
        },
        {
            "key": "CLAUSOLA_PENALE",
            "text": "[INSERIRE penale per inadempimento — revisione legale obbligatoria]",
        },
        {
            "key": "CLAUSOLA_PROPRIETÀ",
            "text": "[INSERIRE dichiarazioni sullo stato di fatto e di diritto dell'immobile]",
        },
        {
            "key": "CLAUSOLA_PRIVACY",
            "text": "[INSERIRE informativa trattamento dati personali]",
        },
        {
            "key": "FIRMA",
            "text": "[FIRMA LEGALE RAPPRESENTANTE + TIMBRO SOCIETÀ]",
        },
    ]

    return OfferLetterResponse(
        commercial_text=commercial,
        legal_placeholders=placeholders,
        disclaimer=DISCLAIMER,
    )

File: apps/web/src/lib/analytics-engine/offer-generator.ts
const DISCLAIMER =
  "BOZZA GENERATA AUTOMATICAMENTE — Non costituisce consulenza legale o fiscale. " +
  "Revisione obbligatoria da parte di avvocato/notaio prima dell'invio.";

const STRATEGY_LABELS: Record<string, string> = {
  fix_flip: "acquisto con finalità di riqualificazione e successiva vendita",
  buy_renovate_rent: "acquisto con ristrutturazione e locazione",
  buy_hold_sell: "acquisto con gestione patrimoniale e exit programmata",
};

export interface OfferLetterRequest {
  property_address: string;
  offered_price: number;
  asking_price: number;
  strategy: string;
  key_points?: string[];
  requested_documents?: string[];
  closing_days?: number;
  deposit_pct?: number;
}

export interface OfferLetterResponse {
  commercial_text: string;
  legal_placeholders: Array<{ key: string; text: string }>;
  disclaimer: string;
}

function formatMoney(value: number): string {
  return value.toLocaleString("it-IT", { minimumFractionDigits: 2, maximumFractionDigits: 2 });
}

export function generateOfferLetter(req: OfferLetterRequest): OfferLetterResponse {
  const askingPrice = req.asking_price;
  const offeredPrice = req.offered_price;
  const discount =
    askingPrice > 0 ? ((askingPrice - offeredPrice) / askingPrice) * 100 : 0;
  const depositPct = req.deposit_pct ?? 0.05;
  const deposit = offeredPrice * depositPct;
  const closingDays = req.closing_days ?? 60;

  const pointsText =
    req.key_points && req.key_points.length > 0
      ? req.key_points.map((p) => `  • ${p}`).join("\n")
      : "  • Prezzo coerente con analisi di mercato e stato dell'immobile\n" +
        "  • Tempistiche definite per perizia e accesso tecnico";

  const docs =
    req.requested_documents ?? [
      "Visura catastale",
      "Planimetria catastale",
      "Atto di provenienza",
      "APE",
      "Conformità urbanistica",
    ];
  const docsText = docs.map((d) => `  • ${d}`).join("\n");

  const commercial = `PROPOSTA DI ACQUISTO — BOZZA COMMERCIALE

Oggetto: Immobile sito in ${req.property_address}

Gentile Venditore,

con la presente formuliamo proposta irrevocabile per ${closingDays} giorni
per l'acquisto dell'immobile in oggetto, alle seguenti condizioni economiche:

  Prezzo richiesto:     € ${formatMoney(askingPrice)}
  Prezzo offerto:       € ${formatMoney(offeredPrice)} (sconto circa ${discount.toFixed(1)}%)
  Caparra confirmatoria: € ${formatMoney(deposit)} (${(depositPct * 100).toFixed(0)}% del prezzo offerto)

Finalità dell'acquisto: ${STRATEGY_LABELS[req.strategy] ?? "investimento immobiliare"}.

Motivazione sintetica:
${pointsText}

Documentazione richiesta prima del rogito:
${docsText}

Tempistiche proposte:
  • Accettazione proposta: entro ${Math.floor(closingDays / 3)} giorni
  • Perizia tecnica e verifiche documentali: entro ${Math.floor(closingDays / 2)} giorni
  • Rogito: entro ${closingDays} giorni dall'accettazione

Restiamo a disposizione per un sopralluogo tecnico e per approfondire
ogni aspetto dell'operazione.

Cordiali saluti,
[SOCIETÀ ACQUIRENTE — da compilare]
`;

  return {
    commercial_text: commercial,
    legal_placeholders: [
      {
        key: "CLAUSOLA_SOSPENSIVA",
        text: "[INSERIRE clausole sospensive: mutuo, conformità urbanistica, assenza vincoli]",
      },
      {
        key: "CLAUSOLA_PENALE",
        text: "[INSERIRE penale per inadempimento — revisione legale obbligatoria]",
      },
      {
        key: "CLAUSOLA_PROPRIETÀ",
        text: "[INSERIRE dichiarazioni sullo stato di fatto e di diritto dell'immobile]",
      },
      {
        key: "CLAUSOLA_PRIVACY",
        text: "[INSERIRE informativa trattamento dati personali]",
      },
      {
        key: "FIRMA",
        text: "[FIRMA LEGALE RAPPRESENTANTE + TIMBRO SOCIETÀ]",
      },
    ],
    disclaimer: DISCLAIMER,
  };
}
